src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -740,27 +741,37 @@ s2a_decrypt_status s2a_decrypt_record(
     return decrypt_status;
   }
 
+  grpc_status_code increment_status =
+      increment_sequence(crypter->in_connection);
+  if (increment_status != GRPC_STATUS_OK) {
+    return INTERNAL_ERROR;
+  }
+
   uint8_t* plaintext = reinterpret_cast<uint8_t*>(unprotected_vec.iov_base);
-  /** Search for the nonzero trailing byte, which encodes the record type and is
-   *  appended to the plaintext. **/
+  /** At this point, the |s2a_decrypt_payload| method has written
+   *  |*bytes_written| bytes to |plaintext|, and these bytes are of the form
+   *    (ciphertext) + (record type byte) + (trailing zeros).
+   *  These trailing zeros should be ignored, so we will search from one end of
+   *  the |plaintext| buffer until we find the first nonzero trailing byte,
+   *  which must be the record type.
+   *
+   *  Note that this TLS 1.3 implementation does not add padding by zeros when
+   *  constructing a TLS 1.3 record; nonetheless, |s2a_decrypt_payload| must be
+   *  able to parse a TLS 1.3 record that does have padding by zeros. **/
   size_t i;
-  for (i = *bytes_written - 1; i < *bytes_written; i--) {
+  for (i = *bytes_written - 1; i >= 0; i--) {
     if (plaintext[i] != 0) {
       break;
     }
   }
-  if (i >= *bytes_written) {
+  if (i < 0) {
     *error_details = gpr_strdup(kS2ARecordInvalidFormat);
     return INVALID_RECORD;
   }
   uint8_t record_type = plaintext[i];
-  *bytes_written -= 1;
-
-  grpc_status_code increment_status =
-      increment_sequence(crypter->in_connection);
-  if (increment_status != GRPC_STATUS_OK) {
-    return INTERNAL_ERROR;
-  }
+  /** The plaintext only occupies the first i bytes of the |plaintext| buffer,
+   *  so |bytes_written| must be updated accordingly. **/
+  *bytes_written = i;
```

Comment:
Should this be `i+1`? Otherwise, we're not counting the record_type in the count.